### ISSUE TITLE
What's the effect of loss_adv_pred

### ISSUE THREAD

**Issue by zhangyuygss** [hfslyc/AdvSemiSeg#2]
In line 371 of script **train.py**:
`loss = loss_seg + args.lambda_adv_pred * loss_adv_pred`

It seems to me that the `loss_adv_pred` part doesn't do anything for the training, since you have set `require_gradient` for the discriminator to False in line 309-311:

`    # don't accumulate grads in D`
`    for param in model_D.parameters():`
 `       param.requires_grad = False`

Am I thinking it wrong? If so, could you please tell me about it?

**Comment by hfslyc** [comment]
We only fix D when training segmentation network and reset it when training D.

Please see https://github.com/hfslyc/AdvSemiSeg/blob/master/train.py#L380-L384

**Comment by zhangyuygss** [comment]
I understand what you mean, what I mean is that the loss_adv_pred in line 371 seems useless for the training since nothing goes backward through it.

**Comment by hfslyc** [comment]
the loss variable is still connected to segmentation network through D. so the backward will not affect D but will backprop to G(segmentation).

**Comment by John1231983** [comment]
Hi. You used
```
for param in model_D.parameters():
                param.requires_grad = False

loss = loss_seg + args.lambda_adv_pred * loss_adv_pred

            # proper normalization
            loss = loss/args.iter_size
            loss.backward()
```

Does it same as the bellow?
```
loss = loss_seg + args.lambda_adv_pred * loss_adv_pred

            # proper normalization
            loss = loss/args.iter_size
            loss_seg.backward()
```
By delete the first two lines and change `loss.backward()` to `loss_seg.backward()` because `loss_adv_pred` does not do backward() so D not trained.
